omf-python/src/file/reader.rs

```diff
@@ -132,84 +166,235 @@ impl PyReader {
+        ))
     }
 
     /// Read an Index array.
-    pub fn array_indices(&self, array: &PyIndexArray) -> PyResult<Vec<Option<u32>>> {
-        self.0
+    pub fn array_indices<'py>(
+        &self,
+        py: Python<'py>,
+        array: &PyIndexArray,
+    ) -> PyResult<Bound<'py, PyArray1<u32>>> {
+        let (indices, mask): (Vec<_>, Vec<_>) = self
+            .0
             .array_indices(&array.0)
             .map_err(OmfException::py_err)?
-            .collect::<Result<Vec<Option<u32>>, _>>()
-            .map_err(OmfException::py_err)
+            .process_results(|r| r.map(|i| (i.unwrap_or_default(), i.is_none())).unzip())
+            .map_err(OmfException::py_err)?;
+
+        Ok(masked_array(
+            py,
+            PyArray::from_owned_array_bound(py, indices.into()),
+            PyArray::from_owned_array_bound(py, mask.into()),
+        )?
+        .downcast_into()?)
     }
 
     /// Read a Triangle array.
-    pub fn array_triangles(&self, array: &PyTriangleArray) -> PyResult<Vec<[u32; 3]>> {
-        self.0
+    pub fn array_triangles<'py>(
+        &self,
+        py: Python<'py>,
+        array: &PyTriangleArray,
+    ) -> PyResult<Bound<'py, PyArray2<u32>>> {
+        let triangles: Vec<_> = self
+            .0
             .array_triangles(&array.0)
             .map_err(OmfException::py_err)?
-            .collect::<Result<Vec<_>, _>>()
-            .map_err(OmfException::py_err)
+            .collect::<Result<_, _>>()
+            .map_err(OmfException::py_err)?;
+
+        let shape = (triangles.len(), 3);
+
+        // TODO: remove unwrap
+
+        Ok(PyArray::from_owned_array_bound(
+            py,
+            Array::from_shape_vec(shape, triangles.into_flattened()).unwrap(),
+        ))
     }
 
     /// Read a Color array.
-    pub fn array_color(&self, array: &PyColorArray) -> PyResult<Vec<Option<Color>>> {
-        self.0
+    pub fn array_color<'py>(
+        &self,
+        py: Python<'py>,
+        array: &PyColorArray,
+    ) -> PyResult<Bound<'py, PyArray2<u8>>> {
+        let (colors, mask): (Vec<_>, Vec<_>) = self
+            .0
             .array_colors(&array.0)
             .map_err(OmfException::py_err)?
-            .collect::<Result<Vec<_>, _>>()
-            .map_err(OmfException::py_err)
+            .process_results(|r| r.map(|c| (c.unwrap_or_default(), [c.is_none(); 4])).unzip())
+            .map_err(OmfException::py_err)?;
+
+        let shape = (colors.len(), 4);
+        // TODO: remove unwrap
+        let colors = PyArray::from_owned_array_bound(
+            py,
+            Array::from_shape_vec(shape, colors.into_flattened()).unwrap(),
+        );
+
+        let shape = (mask.len(), 4);
+        // TODO: remove unwrap
+        let mask = PyArray::from_owned_array_bound(
+            py,
+            Array::from_shape_vec(shape, mask.into_flattened()).unwrap(),
+        );
+
+        Ok(masked_array(py, colors, mask)?.downcast_into()?)
```

Comment:
I suspect that a masked array here won't be all that useful to the caller. NumPy masked arrays mask each element, but for this and similar arrays from OMF it's each row that can be null or not. In my experience doing any processing with a masked array like this is a pain, as operations on rows didn't skip null rows like I expected them to.

This will be of a problem for writing too, where the Python bindings would need to check that masked arrays are following the right pattern.

Returning and accepting an `(array, mask)` tuple might be more useful. For colors the array would have shape (N, 4) and the mask just (N,). The caller could still create a masked array if they want one, or compress the colors with `array.compress(mask, axis=0)` which is a pain with a 2D masked array.

It's been a while since I looked at this though, and I can't remember all the details. Maybe try out a few operations like mapping RGB to sRGB or HSV and see if it works well. This only affects colors and vectors in OMF, but I would prefer consistency across all nullable array types.

Reply:
Okay, returning a pair of numpy arrays makes sense.